<launch>

  <arg name="FRAME_ID" />

  <arg name="PROCESS" default="true" />
  <arg name="STATIC_SCENE" default="false" />
  <arg name="OCCUPANCY" default="false" />

  <arg name="NAMESPACE" value="/camera/with_occupancy" if="$(arg OCCUPANCY)" />
  <arg name="NAMESPACE" value="/camera/without_occupancy" unless="$(arg OCCUPANCY)" />

  <group ns="$(arg NAMESPACE)">

    <node name="singleview_3d_pose_estimation"
          pkg="morefusion_ycb_video" type="singleview_3d_pose_estimation.py"
          clear_params="true"
          output="screen" if="$(arg PROCESS)">
      <remap from="~input/camera_info" to="/camera/color/camera_info" />
      <remap from="~input/rgb" to="/camera/color/image_rect_color" />
      <remap from="~input/depth" to="/camera/aligned_depth_to_color/image_raw" />
      <!-- OCCUPANCY: false -->
      <remap from="~input/label_ins" to="/camera/mask_rcnn_instance_segmentation/output/label_ins" unless="$(arg OCCUPANCY)" />
      <remap from="~input/class" to="/camera/mask_rcnn_instance_segmentation/output/class" unless="$(arg OCCUPANCY)" />
      <!-- OCCUPANCY: true -->
      <remap from="~input/label_ins" to="/camera/octomap_server/output/label_rendered" if="$(arg OCCUPANCY)" />
      <remap from="~input/class" to="/camera/octomap_server/output/class" if="$(arg OCCUPANCY)" />
      <remap from="~input/grids_noentry" to="/camera/octomap_server/output/grids_noentry" if="$(arg OCCUPANCY)" />
      <rosparam subst_value="true">
        with_occupancy: $(arg OCCUPANCY)
      </rosparam>
    </node>
    <node name="object_poses_to_markers"
          pkg="morefusion_ycb_video" type="object_poses_to_markers.py"
          clear_params="true">
      <remap from="~input" to="singleview_3d_pose_estimation/output" />
    </node>
    <node name="render_mesh_markers"
          pkg="morefusion_ycb_video" type="render_mesh_markers.py"
          clear_params="true">
      <remap from="~input/camera_info" to="/camera/color/camera_info" />
      <remap from="~input/markers" to="object_poses_to_markers/output" />
    </node>
    <group ns="render_mesh_markers">
      <node name="tile_image"
            pkg="morefusion_ycb_video" type="tile_image.py"
            clear_params="true">
        <rosparam subst_value="true">
          input_topics:
            - /camera/color/image_rect_color
            - output
          queue_size: 100
          approximate_sync: false
        </rosparam>
      </node>
    </group>

    <group if="$(arg STATIC_SCENE)">
      <node name="object_mapping"
            pkg="morefusion_ycb_video" type="object_mapping.py"
            clear_params="true" if="$(arg PROCESS)">
        <remap from="~input/poses" to="singleview_3d_pose_estimation/output" />
        <remap from="~input/grids" to="/camera/octomap_server/output/grids" />
        <rosparam subst_value="true">
          frame_id: $(arg FRAME_ID)
        </rosparam>
      </node>
      <group ns="object_mapping">
        <node name="grids_to_mesh_markers"
              pkg="morefusion_ycb_video" type="voxel_grids_to_mesh_markers.py"
              clear_params="true">
          <remap from="~input" to="output/grids" />
        </node>
        <node name="grids_to_markers"
              pkg="morefusion_ycb_video" type="voxel_grids_to_markers.py"
              clear_params="true">
          <remap from="~input" to="output/grids" />
          <rosparam subst_value="true">
            show_bbox: false
          </rosparam>
        </node>
        <node name="object_poses_to_markers"
              pkg="morefusion_ycb_video" type="object_poses_to_markers.py"
              clear_params="true">
          <remap from="~input" to="output/poses" />
        </node>
        <node name="render_mesh_markers"
              pkg="morefusion_ycb_video" type="render_mesh_markers.py"
              clear_params="true">
          <remap from="~input/camera_info" to="/camera/color/camera_info" />
          <remap from="~input/markers" to="object_poses_to_markers/output" />
        </node>
        <group ns="render_mesh_markers">
          <node name="tile_image"
                pkg="morefusion_ycb_video" type="tile_image.py"
                clear_params="true">
            <rosparam subst_value="true">
              input_topics:
                - /camera/color/image_rect_color
                - output
              queue_size: 100
              approximate_sync: false
            </rosparam>
          </node>
        </group>
      </group>
    </group>

    <group if="$(arg OCCUPANCY)">
      <node name="collision_based_pose_refinement"
            pkg="morefusion_ycb_video" type="collision_based_pose_refinement.py"
            clear_params="true"
            output="screen" if="$(arg PROCESS)">
        <remap from="~input/poses" to="singleview_3d_pose_estimation/output" />
        <remap from="~input/grids" to="/camera/octomap_server/output/grids" />
        <remap from="~input/grids_noentry" to="/camera/octomap_server/output/grids_noentry" />
      </node>
      <group ns="collision_based_pose_refinement">
        <node name="object_poses_to_markers_debug"
              pkg="morefusion_ycb_video" type="object_poses_to_markers.py"
              clear_params="true">
          <remap from="~input" to="debug" />
        </node>
        <node name="object_poses_to_markers"
              pkg="morefusion_ycb_video" type="object_poses_to_markers.py"
              clear_params="true">
          <remap from="~input" to="output" />
        </node>
        <node name="render_mesh_markers"
              pkg="morefusion_ycb_video" type="render_mesh_markers.py"
              clear_params="true">
          <remap from="~input/camera_info" to="/camera/color/camera_info" />
          <remap from="~input/markers" to="object_poses_to_markers/output" />
        </node>
        <group ns="render_mesh_markers">
          <node name="tile_image"
                pkg="morefusion_ycb_video" type="tile_image.py"
                clear_params="true">
            <rosparam subst_value="true">
              input_topics:
                - /camera/color/image_rect_color
                - output
              queue_size: 100
              approximate_sync: false
            </rosparam>
          </node>
        </group>
        <group if="$(arg STATIC_SCENE)">
          <node name="object_mapping"
                pkg="morefusion_ycb_video" type="object_mapping.py"
                clear_params="true" if="$(arg PROCESS)">
            <remap from="~input/poses" to="output" />
            <remap from="~input/grids" to="/camera/octomap_server/output/grids" />
            <rosparam subst_value="true">
              frame_id: $(arg FRAME_ID)
            </rosparam>
          </node>
          <group ns="object_mapping">
            <node name="grids_to_mesh_markers"
                  pkg="morefusion_ycb_video" type="voxel_grids_to_mesh_markers.py"
                  clear_params="true">
              <remap from="~input" to="output/grids" />
            </node>
            <node name="grids_to_markers"
                  pkg="morefusion_ycb_video" type="voxel_grids_to_markers.py"
                  clear_params="true">
              <remap from="~input" to="output/grids" />
              <rosparam subst_value="true">
                show_bbox: false
              </rosparam>
            </node>
            <node name="object_poses_to_markers"
                  pkg="morefusion_ycb_video" type="object_poses_to_markers.py"
                  clear_params="true">
              <remap from="~input" to="output/poses" />
            </node>
            <node name="render_mesh_markers"
                  pkg="morefusion_ycb_video" type="render_mesh_markers.py"
                  clear_params="true">
              <remap from="~input/camera_info" to="/camera/color/camera_info" />
              <remap from="~input/markers" to="object_poses_to_markers/output" />
            </node>
            <group ns="render_mesh_markers">
              <node name="tile_image"
                    pkg="morefusion_ycb_video" type="tile_image.py"
                    clear_params="true">
                <rosparam subst_value="true">
                  input_topics:
                    - /camera/color/image_rect_color
                    - output
                  queue_size: 100
                  approximate_sync: false
                </rosparam>
              </node>
            </group>
          </group>
        </group>  <!-- if="$(arg STATIC_SCENE)" -->
      </group>  <!-- ns="collision_based_pose_refinement" -->
    </group>  <!-- if="$(arg OCCUPANCY)" -->

  </group>

</launch>
